### PR TITLE
Issue 124 better config (resolves #124, resolves #217, resolves #221)

### LINKED ISSUE #124: Change config to proper class

### LINKED ISSUE #217: Options are overwritten on restart
When restarting a toil workflow the all options are overwritten with what the user specifies + defaults. It would be preferable if only the options that the user specifies be overwritten

### LINKED ISSUE #221: Remove big batch system

### PR DESCRIPTION
**Pull request by benedictpaten**
This addresses issues: #221 #124 #217

### CODE REVIEW

**Comment by benedictpaten** [comment]
I think the tests are stuck in a loop, what do we do in this case?

**Comment by hannes-ucsc** [comment]
Branch name: check. Issue title: nope. Issue number needs to mentioned as "(resolves #124)".

Re hang: I created a new slave image and since then the agent hangs on launch. 

https://github.com/BD2KGenomics/cgcloud/issues/43

For workaround, see above issue.

File: src/toil/common.py [hunk 1]
```diff
@@ -72,87 +46,192 @@ def toilPackageDirPath():
...
+        h2b = lambda x : human2bytes(str(x)) #Function to parse integer from string expressed in different formats
+        
+        def iC(minValue, maxValue=sys.maxint):
+            #Returns function to check the a parameter is in a valid range
+            def f(x):
+                assert x >= minValue and x < maxValue
+            return f
+        
+        #Core options
+        setOption("jobStore", parsingFn=lambda x : os.path.abspath(x)
```
Comment (hannes-ucsc) [inline src/toil/common.py]: parsingFn=os.path.abspath

File: src/toil/common.py [hunk 2]
```diff
@@ -160,50 +239,31 @@ def _addOptions(addGroupFn, defaultStr):
     addOptionFn = addGroupFn("toil options for rescuing/killing/restarting jobs", \
             "The options for jobs that either run too long/fail or get lost \
             (some batch systems have issues!)")
-    addOptionFn("--retryCount", dest="retryCount", default=0,
+    addOptionFn("--retryCount", dest="retryCount", default=None,
```
Comment (hannes-ucsc) [inline src/toil/common.py]: None is the default default.

**Comment by hannes-ucsc** [comment]
Tests stuck in infinite loop:

Mesos executor stderr:
```
DEBUG:toil.batchSystems.mesos.executor:Starting executor
I0817 17:06:20.035890  4045 exec.cpp:132] Version: 0.23.0
I0817 17:06:20.038673  4048 exec.cpp:206] Executor registered on slave 20150817-170619-16777343-5050-4020-S0
INFO:toil.batchSystems.mesos.executor:Registered with framework
DEBUG:toil.batchSystems.mesos.executor:Running task 0
DEBUG:toil.batchSystems.mesos.executor:Sending status update ...
DEBUG:toil.batchSystems.mesos.executor:sent stats message
DEBUG:toil.batchSystems.mesos.executor:Sent status update
DEBUG:toil.batchSystems.mesos.executor:Invoking command: '/mnt/ephemeral/workspace/toil/venv/bin/python2.7 -E /mnt/ephemeral/workspace/toil/src/toil/worker.py ./toil 7/f/jobjIomht'
Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil/src/toil/worker.py", line 434, in <module>
    main()
  File "/mnt/ephemeral/workspace/toil/src/toil/worker.py", line 125, in main
    jobStore = loadJobStore(jobStoreString)
  File "/mnt/ephemeral/workspace/toil/src/toil/common.py", line 376, in loadJobStore
    return FileJobStore( jobStoreArgs, config=config )
  File "/mnt/ephemeral/workspace/toil/src/toil/jobStores/fileJobStore.py", line 35, in __init__
    self._checkJobStoreCreation(config != None, os.path.exists(self.jobStoreDir), self.jobStoreDir)
  File "/mnt/ephemeral/workspace/toil/src/toil/jobStores/abstractJobStore.py", line 80, in _checkJobStoreCreation
    "is nothing to restart." % jobStoreString)
toil.jobStores.abstractJobStore.JobStoreCreationException: The job store '/tmp/mesos/slaves/20150817-170619-16777343-5050-4020-S0/frameworks/20150817-170619-16777343-5050-4020-0000/executors/badExecutor/
runs/aced20a7-4aef-4c57-b458-0575c2f167d8/toil' does not exist, so there is nothing to restart.
DEBUG:toil.batchSystems.mesos.executor:Sending status update ...
DEBUG:toil.batchSystems.mesos.executor:Sent status update
```

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
The Jenkins build skipped all AWS-related tests.

b3847affbdf221a2189c2c28b4b1ff8f41a49c80 needs to be merged from master in order to fix that.

**Comment by cket** [comment]
The mesos problem is caused by a relative path to the toil directory being passed with the worker (Python -E worker.py ./toil  etc...) The executor does not operate in the same directory as the master, so ./toil does not exist in it's directory. This issue came up before and was/is fixed on master by making this /.toil into an absolute path.

**Comment by benedictpaten** [comment]
I pull in the master this morning?

**Comment by benedictpaten** [comment]
Okay, I know what I did wrong, CJ helped me, it's running now

**Comment by benedictpaten** [comment]
Okay, the only test that is reported to have failed is pytest, and I think that's a CI issue, hannes?

**Comment by benedictpaten** [comment]
Also, looking at the test report it seems not all tests are being run? For example, toil.test.utils.utilTest

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
```ENOENT: [No such file or directory]: getcwd()```

happens when the current directory has been deleted. The issue is specific to this branch, not CI in general since master is building ok, also with py.test.

**Comment by benedictpaten** [comment]
Look like we are finally all good!

**Pull request merged**